Add custom elements to HTML

domenic opened the pull request:
This upstreams custom elements into the HTML Standard, instead of having
them be in a separate spec that monkeypatches HTML extensively. This
builds on previous work in https://github.com/whatwg/dom/pull/204.

See https://rawgit.com/w3c/webcomponents/15a203c8393aef0df7223ab1d43406aa11a7e71e/spec/custom/index.html
for the source material, in particular the sections labeled "HTML+". All
changes while upstreaming were editorial in nature.

---

Review appreciated, especially in the form of pushing fixups to the branch.

annevk commented on images/content-venn.svg:
```diff
@@ -108,6 +108,7 @@
      <li><code>var</code></li>
      <li><code>video</code></li>
      <li><code>wbr</code></li>
+     <li><span>custom tags</span></li>
```
The plan is to rename this in a follow up commit?

domenic replied:
Yeah, I guess, if we ever figure something out.

annevk replied:
Well I think we need to. "tags" is just super wrong.

annevk commented:
I should have asked this earlier, where exactly does NewTarget come from? Is that worthy of a note or a cross-reference?

domenic commented:
> I should have asked this earlier, where exactly does NewTarget come from? Is that worthy of a note or a cross-reference?

I think it would be ideal to cross-reference, but ES does not have a definition for it; instead as far as I can tell it's just an ambient floating thing all ES algorithms get. I'll file an ES bug.

domenic commented on source:
```diff
@@ -9569,8 +9569,18 @@ interface <dfn>HTMLUnknownElement</dfn> : <span>HTMLElement</span> { };</pre>
+   <li><p>If <var>name</var> is <code>blink</code>, <code>bgsound</code>, <code>isindex</code>,
+   <code>multicol</code>, <code>nextid</code>, or <code>spacer</code>, return
+   <code>HTMLUnknownElement</code>.</p></li>
+   <!-- These have to be explicitly listed because technically we define these elements in HTML,
```
This helps with https://www.w3.org/Bugs/Public/show_bug.cgi?id=27877, but probably doesn't completely take care of it, as basefont might still be confusing.

domenic commented on source:
```diff
@@ -9569,8 +9569,18 @@ interface <dfn>HTMLUnknownElement</dfn> : <span>HTMLElement</span> { };</pre>
-   to the tag name <var>name</var>, return that interface.</p></li>
+   to the local name <var>name</var>, return that interface.</p></li>
+
+   <li><p>If this specifications defines the element type corresponding to the local name
```
xref element type

domenic replied:
s/specifications/specification

domenic commented on source:
```diff
@@ -9569,8 +9569,18 @@ interface <dfn>HTMLUnknownElement</dfn> : <span>HTMLElement</span> { };</pre>
+
+   <li><p>If this specifications defines the element type corresponding to the local name
+   <var>name</var>, but not an interface, return <code>HTMLElement</code>.</p></li>
+   <!-- Otherwise, e.g., <listing> would not be covered. -->
```
Hmm maybe this covers the basefont case? A note listing the tags this applies to might be good.

annevk replied:
If we start doing that we should maybe just go all the way and use a table.

domenic replied:
Hmm. Could work, I suppose.

domenic commented on source:
```diff
@@ -66841,11 +66851,12 @@ dictionary <dfn>ElementRegistrationOptions</dfn> {
    <dd>Defines a new <span>custom element</span>, mapping the given name to the given constructor as
-   a <span>type extension</span> for the supplied base tag. A <code>NotSupportedError</code> will be
-   thrown upon trying to extend a <span>custom element</span> or an unknown element.</dd>
+   a <span>type extension</span> for the supplied <var>base</var> element. A
```
I don't like this because it implies _base_ is an element (i.e. an object), instead of a tag name (a string).

annevk replied:
Well, it's an extension for an element represented by a local name (not a tag name).

domenic replied:
Maybe baseLocalName then. I thought tag name was the more author-friendly way of saying that.

annevk replied:
The HTML specification uses "tag" only for syntax and parsing (and for xml:space, but that seems like a bug).

annevk replied:
That's why "custom tags" is so upsetting.

domenic replied:
Well, the syntax is custom.

annevk replied:
Except that it's mostly API-driven, which don't deal with tags. Tags have angle brackets.

domenic replied:
Yeah, but the tags show up in the source as custom too. "Custom tags" in my mind is a shortening of "Element types that show up in HTML source as custom tags."

annevk replied:
But the source is just input (and serialization). What's key is the data model. Everything else is defined in terms of the data model. Why would we deviate here?

domenic replied:
Because we haven't come up with good names yet. Anyway, we can rename custom tags or whatever. My point in these line comments is that you should not conflate elements with strings, whether you want to call those strings local names or tag names or names or.... Strings is what we need to describe in these lines.

domenic commented on source:
```diff
@@ -66882,7 +66893,7 @@ dictionary <dfn>ElementRegistrationOptions</dfn> {
 
      <li><p>If the <span>element interface</span> for <var>extends</var> and the <span>HTML
-     namespace</span> is <code>HTMLUnknownElement</code> (i.e. if <var>extends</var> does not
+     namespace</span> is <code>HTMLUnknownElement</code> (e.g., if <var>extends</var> does not
```
I think this should be `i.e.` (with a comma if you want); the parenthetical is not an example of what could happen, but instead exactly the reason why this would happen.

annevk replied:
Not now I listed other reasons for getting HTMLUnknownElement.

domenic replied:
Ah, right, of course; thanks.

domenic commented:
I pushed fixes addressing the issues I noted on @annevk's commit. One interesting issue it raised is defining element interface for obsolete elements better. I will open a new issue on that.

domenic commented:
Should "Core concepts" be named "Processing model"? Maybe not, since so much of the processing model is defined elsewhere...

annevk commented on source:
```diff
@@ -9574,26 +9574,28 @@ interface <dfn>HTMLUnknownElement</dfn> : <span>HTMLElement</span> { };</pre>
    <code>multicol</code>, <code>nextid</code>, or <code>spacer</code>, return
    <code>HTMLUnknownElement</code>.</p></li>
    <!-- These have to be explicitly listed because technically we define these elements in HTML,
         albeit as obsolete. -->
```
This comment is obsolete with this commit, no?

domenic replied:
Yeah good catch

annevk commented:
Would you mind if I added ", then" to conditionals?

annevk commented:
Although with all the don't squash that might get a little labor intensive...

annevk commented on source:
```diff
@@ -66858,6 +66858,7 @@ console.log(plasticButton2.getAttribute("is")); // will output "plastic-button"<
 
   <pre class="idl">interface <dfn>CustomElementsRegistry</dfn> {
   void <span data-x="dom-CustomElementsRegistry-define">define</span>(DOMString name, Function constructor, optional ElementRegistrationOptions options);
+  any <span data-x="dom-CustomElementsRegistry-get">get</span>(DOMString name);
```
s/any/Function/

domenic commented:
You can add it, it's fine. Maybe don't add it to the sections touched in "don't squash" and I can do those myself when I rebase and rearrange commits.

annevk commented:
I think it's all good now. I have fixed up the commits to add "then" and removed "don't squash" from the titles. Admittedly I have relied in part on my earlier review of the algorithms. Since I assume you just moved them around.

zcorpan commented on source:
```diff
@@ -67296,6 +67305,18 @@ dictionary <dfn>ElementRegistrationOptions</dfn> {
+   data-x="concept-custom-element-definition-name">name</span> <var>name</var>, then return that
+   entry's <span data-x="concept-custom-element-definition-constructor">constructor</span>.</p></li>
+
+   <li><p>Otherwise, return undefined.</p></li>
```
I think it's more common to return null from a method. Why undefined?

annevk replied:
I noticed that. I suspect @domenic wants to mimic Map semantics.

domenic replied:
Yes, exactly, map-like get() methods need to return undefined.

zcorpan commented:
There needs to be a normative statement saying that `<html is>` is invalid, e.g. "The `is` attribute must not be specified."

zcorpan commented:
> "rationalise the platform,"

Move the comma to the right of the " per Hixie English.

zcorpan commented:
Autonomous custom elements should also not allow the `is` attribute, right?

zcorpan commented:
```diff
+  attributeChangedCallback() {
+    // only is called for the disabled attribute due to observedAttributes
+    if (this.disabled) {
+      this.setAttribute("tabindex", "-1");
```

Shouldn't this use `removeAttribute`?

zcorpan commented:
Do we want to allow any attribute on customized built-in elements? In particular those in the obsolete section? Note that `embed` allows any attribute *except* those that are obsolete (e.g. `name`).

zcorpan commented:
```diff
+   <dt><var>window</var> . <code data-x="dom-window-customElements">customElements</code> . <code
+   data-x="dom-CustomElementsRegistry-define">define</code>(<var>name</var>,
+   <var>constructor</var>)</dt>
```
These typically have a `subdfn` attribute (which I think is for developers.whatwg.org?).

zcorpan commented:
(Reached the end of the diff; no more comments.)

domenic commented:
>There needs to be a normative statement saying that `<html is>` is invalid, e.g. "The is attribute must not be specified."

Why?

> Autonomous custom elements should also not allow the is attribute, right?

Good catch, thanks.

> Shouldn't this use removeAttribute?

Either would work, I imagine. -1 is more explicit IMO?

> Do we want to allow any attribute on customized built-in elements? In particular those in the obsolete section? Note that embed allows any attribute except those that are obsolete (e.g. name).

Hmm. We probably should? I have no real idea how to phrase this though, or which sections to edit. Suggestions?

> These typically have a subdfn attribute (which I think is for developers.whatwg.org?).

Thanks, will add.

zcorpan commented:
> Either would work, I imagine. -1 is more explicit IMO?

-1 means it's still focusable if you click it. Also the code above uses `removeAttribute` and the prose talks about removing the attribute.

zcorpan commented:
> Why?

The content model is normative, but nothing says the attributes list is. The rest of the spec addresses each attribute individually in prose I believe. So right now it's a statement of fact that's not backed up by anything normative. :-)

annevk commented:
Why can't `is` be specified on `html` anyway? It works, right?

domenic commented:
Yeah, that was my question.

I agree with you now about removeAttribute; thanks for that.

zcorpan commented:
Oh, I don't mind allowing `is` on `html`. Then remove `, except the <code data-x="attr-is">is</code> attribute`.

zcorpan commented:
> Hmm. We probably should? I have no real idea how to phrase this though, or which sections to edit. Suggestions?

How about...

```diff
diff --git a/source b/source
index 0967e3a..a627fa4 100644
--- a/source
+++ b/source
@@ -28002,8 +28002,8 @@ attribute, set the <span>browsing context name</span> of the element's <span>nes
   <p class="note">All attributes in <span>HTML documents</span> get lowercased automatically, so the
   restriction on uppercase letters doesn't affect such documents.</p>
 
-  <p class="note">The four exceptions are to exclude legacy attributes that have side-effects beyond
-  just sending parameters to the <span>plugin</span>.</p>
+  <p class="note">The four exceptions are to exclude <span>obsolete attributes</span> that have
+  side-effects beyond just sending parameters to the <span>plugin</span>.</p>
 
   <div w-nodev>
 
@@ -67011,9 +67011,13 @@ console.log(plasticButton2.getAttribute("is")); // will output "plastic-button"<
 
   <p>An <span>autonomous custom element</span> does not have any special meaning: it
   <span>represents</span> its children. A <span>customized built-in element</span> inherits the
-  semantics of the element that it extends. Both types of <span>custom element</span> may have any
-  non-namespaced attributes that are relevant to the element's functioning, as determined by the
-  element's author.</p>
+  semantics of the element that it extends.</p>
+
+  <p>Any namespace-less attribute, that are relevant to the element's functioning, as determined by
+  the element's author, may be specified on both types of <span data-x="custom element">custom
+  elements</span>, except those that are defined as being <span data-x="obsolete
+  attributes">obsolete</span> for that element, so long as its name is <span>XML-compatible</span>
+  and contains no <span>uppercase ASCII letters</span>.</p>
 
   <p>A <dfn data-export="">valid custom element name</dfn> is a sequence of characters
   <var>name</var> that meets all of the following requirements:</p>
@@ -111655,7 +111659,8 @@ if (s = prompt('What is your name?')) {
 
   <h3>Non-conforming features</h3>
 
-  <p>Elements in the following list are entirely obsolete, and must not be used by authors:</p>
+  <p>Elements in the following list are entirely <dfn data-x="obsolete elements">obsolete</dfn>, and
+  must not be used by authors:</p>
 
   <dl><!-- alphabetical by first element in the group, except CSS goes last -->
 
@@ -111736,8 +111741,8 @@ if (s = prompt('What is your name?')) {
 
   <hr>
 
-  <p>The following attributes are obsolete (though the elements are still part of the language), and
-  must not be used by authors:</p>
+  <p>The following attributes are <dfn data-x="obsolete attributes">obsolete</dfn> (though the
+  elements are still part of the language), and must not be used by authors:</p>
 
   <dl><!-- alphabetical by element then attribute of first item in group, except CSS goes last -->
 
```

Note: I copied the requirements from `embed` that the attribute name be lowercase and XML-compatible.

Additional thoughts:

* The requirements on element names are similar but are defined differently. Editorial but is a bit inconsistent right now.
* Why are any attributes allowed on customized built-in elements? Are they not as problematic for future extensions to the language as any attribute on standard elements?

domenic commented:
> Then remove `, except the <code data-x="attr-is">is</code> attribute`.

Huh, I didn't notice that. I think it must have been intended to go on autonomous built-in elements; I have no idea how it made it to `<html>`.

> Why are any attributes allowed on customized built-in elements? Are they not as problematic for future extensions to the language as any attribute on standard elements?

That's a good point. Let's leave them invalid for now, and address this in a follow-up issue.

domenic commented:
@zcorpan please take a look, made a fixup commit for easier reviewing.

zcorpan commented:
LGTM

annevk commented:
\o/

(Do we want to keep this branch alive as we did in DOM?)

domenic commented:
Yeah, we probably should, and I can also help maintain a similar wiki page.

annevk commented:
Protected.
